nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/Util.java

```diff
@@ -0,0 +1,122 @@
+            final String truststorePass = sslService.getTrustStorePassword();
+            final String truststoreType = sslService.getTrustStoreType();
+
+            KeyStore truststore = KeyStore.getInstance(truststoreType);
+            truststore.load(new FileInputStream(truststoreLocation), truststorePass.toCharArray());
+            trustManagerFactory.init(truststore);
+        }
+
+         /*
+            TrustManagerFactory.getTrustManagers returns a trust manager for each type of trust material. Since we are getting a trust manager factory that uses "X509"
+            as it's trust management algorithm, we are able to grab the first (and thus the most preferred) and use it as our x509 Trust Manager
+            https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/TrustManagerFactory.html#getTrustManagers--
+         */
+        final X509TrustManager x509TrustManager;
+        TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+        if (trustManagers[0] != null) {
+            x509TrustManager = (X509TrustManager) trustManagers[0];
+        } else {
+            throw new IllegalStateException("List of trust managers is null");
+        }
+
+        // if keystore properties were not supplied, the keyManagers array will be null
+        sslContext.init(keyManagers, trustManagerFactory.getTrustManagers(), null);
+
+        final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+        okHttpClientBuilder.sslSocketFactory(sslSocketFactory, x509TrustManager);
+        if (setAsSocketFactory) {
+            okHttpClientBuilder.socketFactory(sslSocketFactory);
+        }
+    }
+
+    public static final String KEY_ACCESS_TOKEN = "access_token";
+    public static final String KEY_REFRESH_TOKEN = "refresh_token";
+    public static final String KEY_EXPIRES = "expires_in";
+    public static final String KEY_TOKEN_TYPE = "token_type";
+    public static final String KEY_SCOPE = "scope";
+
+    public static AccessToken parseTokenResponse(String rawResponse) {
+        try {
+            Map<String, Object> parsed = new ObjectMapper().readValue(rawResponse, Map.class);
+            String accessToken = (String)parsed.get(KEY_ACCESS_TOKEN);
+            String refreshToken = (String)parsed.get(KEY_REFRESH_TOKEN);
+            Integer expires = (Integer)parsed.get(KEY_EXPIRES);
+            String tokenType = (String)parsed.get(KEY_TOKEN_TYPE);
+            String scope = (String)parsed.get(KEY_SCOPE);
+
+            if (StringUtils.isEmpty(accessToken)) {
+                throw new Exception(String.format("Missing value for %s", KEY_ACCESS_TOKEN));
+            }
+
+            if (StringUtils.isEmpty(tokenType)) {
+                throw new Exception(String.format("Missing value for %s", KEY_TOKEN_TYPE));
+            }
+
+            return new AccessToken(accessToken, refreshToken, tokenType, expires, scope);
+        } catch (Exception ex) {
+            throw new ProcessException(ex);
```

Comment:
Just a question, how do you envision a consuming processor to behave when something like this occurs?

Reply:
Good point. I'll refactor that to be a custom exception. In principal we just halt the execution in our custom bundle that uses this.